Add the `default_with_output_lists` callback plugin

stevekuznetsov opened the pull request:
The `default_with_output_lists` callback plugin replaces a number of
fields like `stdout` and `stderr` with their equivalents, `stdout_lines`
and the like. The plugin also overrides the default output plugin to
always use indented multi-line JSON. Especially when used with `-vv`,
this plugin allows for much more readable output from Ansible.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
Add the `default_with_output_lists` callback plugin

The `default_with_output_lists` callback plugin replaces a number of
fields like `stdout` and `stderr` with their equivalents, `stdout_lines`
and the like. The plugin also overrides the default output plugin to
always use indented multi-line JSON. Especially when used with `-vv`,
this plugin allows for much more readable output from Ansible.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov commented on Dockerfile:
```diff
@@ -56,6 +56,9 @@ RUN mkdir -p /usr/share/ansible $HOME/.ssh $WORK/playbooks/files && \
     ./google-cloud-sdk/install.sh -q --usage-reporting false && \
     echo 'ALL ALL=(ALL) NOPASSWD: ALL' >> /etc/sudoers && \
+    mkdir -p /usr/share/ansible/plugins/callback && \
+    curl -sS https://raw.githubusercontent.com/openshift/origin-ci-tool/master/oct/ansible/oct/callback_plugins/default_with_output_lists.py > /usr/share/ansible/plugins/callback/default_with_output_lists.py && \
```
If you don't care about layers I'll do an `ADD`

stevekuznetsov commented on Dockerfile:
```diff
@@ -56,6 +56,9 @@ RUN mkdir -p /usr/share/ansible $HOME/.ssh $WORK/playbooks/files && \
     echo 'ALL ALL=(ALL) NOPASSWD: ALL' >> /etc/sudoers && \
+    mkdir -p /usr/share/ansible/plugins/callback && \
+    curl -sS https://raw.githubusercontent.com/openshift/origin-ci-tool/master/oct/ansible/oct/callback_plugins/default_with_output_lists.py > /usr/share/ansible/plugins/callback/default_with_output_lists.py && \
+    sed -i 's/^#?stdout_callback/stdout_callback = default_with_output_lists/' /etc/ansible/ansible.cfg && \
```
We can set this at run-time if you want

stevekuznetsov commented:
@openshift-bot you dare question me?!?!?

smarterclayton commented:
[test]

smarterclayton commented:
```
+ sudo -E ../../bin/local.sh ansible-playbook -e openshift_test_repo=https://storage.googleapis.com/origin-ci-test/branch-logs/origin/master/builds/1229/artifacts/rpms playbooks/launch.yaml
Activated service account credentials for: [jenkins-ci-provisioner@openshift-gce-devel.iam.gserviceaccount.com]
ERROR! Attempted to execute "/usr/share/ansible/openshift-ansible-gce/inventory.sh" as inventory script: Inventory script (/usr/share/ansible/openshift-ansible-gce/inventory.sh) had an execution error: error: Inventory configuration failed
 [WARNING]: provided hosts list is empty, only localhost is available
ERROR! Unexpected Exception: cannot import name default
to see the full traceback, use -vvv
```

You should know that the inventory run is also an ansible playbook.

stevekuznetsov committed:
Add the `generate_junit` callback plugin

The `generate_junit` callback plugin generates jUnit XML from Ansible
task execution in order to allow for log aggregators to identify
failures more easily.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov commented:
I don't know what is going on in those tests or what you've pasted -- added in the jUnit plugin as well.

stevekuznetsov commented:
@smarterclayton can we re-trigger tests on this?

stevekuznetsov commented:
@smarterclayton let's try to get this in ... reading your logs is a really unpleasant process today

smarterclayton commented:
Will merge this after I get the fix in

smarterclayton commented:
[test]

smarterclayton commented:
Your PR is still breaking inventory.  Fix it.

stevekuznetsov commented:
Very unclear WTF the error is ... can we get a `-vvv` in there?

stevekuznetsov commented:
Bad regex... should be fine now

stevekuznetsov commented:
OK.... I'm going to need an easier reproducer than using GCS and running that job. Doing a build on my local daemon gets me the plugins in the right place and `/etc/ansible/ansible.cfg` with the right entries.

stevekuznetsov commented:
This is going to need a little bit more work

smarterclayton commented:
Inventory.sh calls ansible.  If ansible that inventory calls is using the callback plugin, it's likely eating the output that inventory needs.

stevekuznetsov commented:
Wow that's amazing

stevekuznetsov commented on Dockerfile:
```diff
@@ -29,7 +29,8 @@ LABEL io.k8s.display-name="OpenShift GCE Install Environment" \
     GOOGLE_CLOUD_SDK_VERSION=130.0.0 \
-    OPENSHIFT_ANSIBLE_TAG=master
+    OPENSHIFT_ANSIBLE_TAG=master \
+    ANSIBLE_JUNIT_DIR=/tmp/openshift/ansible_junit
```
Where do you want the jUnit?

smarterclayton replied:
I have no idea - that location seems fairly weird.

stevekuznetsov replied:
It can be any location as long as you remember to slurp it later

stevekuznetsov commented:
@smarterclayton ok after the nonsense with imports it works --- where do you want the jUnit, and let's get this done

smarterclayton commented:
Conflicts, rebase

stevekuznetsov commented:
@smarterclayton  rebased

stevekuznetsov commented:
@smarterclayton 

>  where do you want the jUnit, and let's get this done

smarterclayton commented:
I don't really care on location I guess.

[test]
